### DISCUSSION TITLE
Daemon/Background service metrics

### DISCUSSION THREAD

**Discussion by dylan-asos** [open-telemetry/opentelemetry-dotnet#4630] [Q&A]
Hi,

We're collecting metrics from our dotnet Web APIs using the Open Telemetry package [OpenTelemetry.Exporter.Prometheus.AspNetCore](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Prometheus.AspNetCore/README.md), which we're initialising at startup with a call to `UseOpenTelemetryPrometheusScrapingEndpoint` to expose Prometheus metrics for scraping - this is deployed on Azure Kubernetes Service. All is good - metrics are exposed on the port we configure in `UseOpenTelemetryPrometheusScrapingEndpoint`  . 

We also have a number of Daemon processing jobs (message processors and other event triggered background jobs) we'd like to expose metrics for scraping. 

Is there a recommended package/approach/example for exporting Prometheus metrics for scraping from a Background worker service? Example skeleton code of the type of app we'd be building.

```csharp
public class Program
{
    public static async Task Main(string[] args)
    {
        var builder = CreateBuilder(args);
        var host = Build(builder);
        await host.RunAsync();
    }

    public static IHostBuilder CreateBuilder(params string[] args)
    {
        var builder = Host.CreateDefaultBuilder(args);
        builder.ConfigureServices(ConfigureServices);
        return builder;
    }

    public static IHost Build(IHostBuilder builder)
    {
        return builder.Build();
    }

    private static void ConfigureServices(HostBuilderContext context, IServiceCollection services)
    {
        services.AddHostedService<Worker>();
    }
}
```

The package [OpenTelemetry.Exporter.Prometheus.HttpListener](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Prometheus.HttpListener/README.md) does what I want, e.g. just call 

```csharp
        builder.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(serviceName));
        builder.AddPrometheusHttpListener(
            options => options.UriPrefixes = new[] {"http://localhost:9464/"});
```

and that would expose that endpoint which could be scraped. However, the statement in the package page about no plans for production & mentioning different packages is making me question if this is the correct approach.

**Comment by paulredert** [comment]
I have the same question. What approach can we follow when creating background worker services? Any news?

**Comment by cijothomas** [comment]
If you are using an Asp.Net Core web app with some background worker in it (i.e same process), then the existing Prometheus.AspNetCore can be used. If you are running a pure worker service which do not by itself run a webserver, then you are limited to using OTLP Exporter only. (unless you want to upgrade it to a webserver, at which point its no diff. than regular asp.net core webhost) The warning in Prometheus.HttpListener is true, and already suggests the approach of using OTLP+collector->prometheus

**Comment by Charlotte-br560** [comment]
Absolutely! To expose Prometheus metrics for scraping from a background worker service, you might consider using the OpenTelemetry.Exporter.Prometheus.HttpListener package. Despite the mention of no plans for production, it could still be a viable approach for your use case.

Here's a simplified example of integrating it into your existing worker service:

`public class Program
{
    public static async Task Main(string[] args)
    {
        var builder = CreateBuilder(args);
        var host = Build(builder);
        await host.RunAsync();
    }

    public static IHostBuilder CreateBuilder(params string[] args)
    {
        var builder = Host.CreateDefaultBuilder(args);
        builder.ConfigureServices(ConfigureServices);
        return builder;
    }

    public static IHost Build(IHostBuilder builder)
    {
        return builder.Build();
    }

    private static void ConfigureServices(HostBuilderContext context, IServiceCollection services)
    {
        services.AddHostedService<Worker>();
        ConfigureOpenTelemetry(services);
    }

    private static void ConfigureOpenTelemetry(IServiceCollection services)
    {
        services.AddOpenTelemetryTracing(builder =>
        {
            // Your other tracing configurations

            builder.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("your-service-name"));

            builder.AddPrometheusHttpListener(options =>
            {
                options.UriPrefixes = new[] { "http://localhost:9464/" };
            });
        });
    }
}
`
This should provide you with the Prometheus metrics endpoint for scraping from your background worker service. Feel free to adjust the configurations based on your specific requirements. Happy coding!
